File: LeetCodeTasks/3011_find-if-array-can-be-sorted/solutions/python3/solution.py
from typing import List

class Solution:
    def canSortArray(self, nums: List[int]) -> bool:
        def count_set_bits(num):
            count = 0
            while num:
                count += num & 1
                num >>= 1
            return count
        
        for i in range(len(nums) - 1):
            if count_set_bits(nums[i]) != count_set_bits(nums[i + 1]):
                return False
        return True

File: LeetCodeTasks/3011_find-if-array-can-be-sorted/solutions/java/solution.java
import java.util.*;

class Solution {
    public boolean canSortArray(int[] nums) {
        List<Integer>[] bitCount = new List[32];
        for (int i = 0; i < 32; i++) {
            bitCount[i] = new ArrayList<>();
        }
        
        for (int num : nums) {
            int count = Integer.bitCount(num);
            bitCount[count].add(num);
        }
        
        for (int i = 0; i < 31; i++) {
            Collections.sort(bitCount[i]);
        }
        
        for (int i = 1; i < nums.length; i++) {
            int count1 = Integer.bitCount(nums[i - 1]);
            int count2 = Integer.bitCount(nums[i]);
            if (count1 != count2) {
                return false;
            }
        }
        
        return true;
    }
}
